csrc/multidevice/utils.cpp

```diff
@@ -96,4 +99,58 @@ bool isResharding(Expr* expr) {
   return !haveDifferentSharding(tv_ref, tvs).empty();
 }
 
+namespace {
+
+void shardAllLike(TensorView* ref, std::vector<TensorView*> tvs) {
+  for (auto tv : tvs) {
+    tv->setDeviceMesh(ref->getDeviceMesh());
+  }
+  scheduler_utils::parallelizeAllLike(ref, tvs, {ParallelType::DIDx});
+}
+
+void reshardBefore(Expr* expr, Fusion* fusion) {
+  NVF_ERROR(
+      expr->outputs().size() == 1,
+      "multi-output expressions are not supported");
+  NVF_ERROR(
+      expr->outputs().at(0)->isA<TensorView>(),
+      "the expression's output is not a TensorView");
+  TensorView* output = expr->outputs().at(0)->as<TensorView>();
+  std::unordered_set<TensorView*> inputs;
+  std::transform(
+      expr->inputs().begin(),
+      expr->inputs().end(),
+      std::inserter(inputs, inputs.end()),
+      [](Val* val) {
+        NVF_ERROR(
+            val->isA<TensorView>(),
+            "the expression's input is not a TensorView");
+        return val->as<TensorView>();
+      });
+  std::vector<TensorView*> new_inputs;
+  for (auto input : haveDifferentSharding(output, inputs)) {
```

Comment:
```suggestion
  TensorView* output = expr->outputs().at(0)->as<TensorView>();
  std::vector<TensorView*> new_inputs;
  for (auto input_tv: ir_utils::filterByType<TensorView>(expr->inputs())) {
  // Why haveDifferentSharding needs to have all tensor inputs? Can it be just the output and one input tensor like this?
  if (haveDifferentSharding(output, input_tv)) {
```

Reply:
Ok for filterByType, I actually used this one before, but wanted to assert that all the inputs are TensorViews (mainly for simplicity, I'm always assuming I am in this case because we have not encountered a situation where it is not the case so far).
But alright I replace as you suggest

> // Why haveDifferentSharding needs to have all tensor inputs? Can it be just the output and one input tensor like this?

Because we need run through all the inputs. Note that `haveDifferentSharding` will return the set of all the inputs which sharding is different than the output

Reply:
> Don't you think it is worth adding those asserts ?

I don't think we have any non-tensor input for tensor ops ATM, but even there's one, we could just ignore non-tensor input as it shouldn't require any communication (or, in other words, we don't support distributing non-tensor values).

> Note that haveDifferentSharding will return the set of all the inputs which sharding is different than the output

I looked at the function, and I'm sorry my previous review was sloppy. I have some concerns, but let's leave it as is for now. I'll open a separate issue.

Reply:
#1656

Reply:
> > Don't you think it is worth adding those asserts ?
> 
> I don't think we have any non-tensor input for tensor ops ATM, but even there's one, we could just ignore non-tensor input as it shouldn't require any communication (or, in other words, we don't support distributing non-tensor values).

ok so I will remove the assert and use filterByType

> > Note that haveDifferentSharding will return the set of all the inputs which sharding is different than the output
> 
> I looked at the function, and I'm sorry my previous review was sloppy. I have some concerns, but let's leave it as is for now. I'll open a separate issue.

Alright I will take a look at the issue

Reply:
Actually I don't think filterByType is suitable here because I want to construct a std::set<TensorView*> to feed the function `haveDifferentSharding`. 

I could, as you suggest, remove the assertion, but I also would rather be extra careful and get an error the day we encounter a case we didn't think about -- in this case if the outputs are not tensors.

So, personally, I wouldn't change anything here. Let me know what you suggest

Reply:
The issue https://github.com/NVIDIA/Fuser/issues/1656 you opened is addressed in a separate PR https://github.com/NVIDIA/Fuser/pull/1674